buttercup/cogs/find.py

```diff
@@ -1,69 +1,238 @@
+
+
+def limit_str(text: str, limit: Optional[int] = None) -> str:
+    """Limit the string to the given length."""
+    if limit is None or len(text) <= limit:
+        return text
+
+    return f"{text[:(limit - 3)]}..."
+
+
+class Find(Cog):
+    def __init__(self, bot: ButtercupBot, blossom_api: BlossomAPI) -> None:
+        """Initialize the Find cog."""
+        self.bot = bot
+        self.blossom_api = blossom_api
+
+    def to_embed(self) -> Embed:  # noqa: C901
+        """Convert the submission to a Discord embed."""
+        color = Color.from_rgb(255, 176, 0)  # Orange
+        status: str = "Unclaimed"
+
+        tr_link = f"[Link]({self.transcription['url']})" if self.transcription else None
+
+        if self.submission["completed_by"]:
+            color = Color.from_rgb(148, 224, 68)  # Green
+
+            if self.author:
+                status = "Completed by {}".format(
+                    DISCORD_USERNAME_LINK.format(self.author["username"])
+                )
+            else:
+                status = "Completed"
+        elif self.submission["claimed_by"]:
+            color = Color.from_rgb(13, 211, 187)  # Cyan
+
+            if self.author:
+                status = "Claimed by {}".format(
+                    DISCORD_USERNAME_LINK.format(self.author["username"])
+                )
+            else:
+                status = "Claimed"
+
+        embed = (
+            Embed(color=color)
+            .add_field(name="Status", value=status)
+            .add_field(
+                name="OCR",
+                value="Yes" if self.submission["has_ocr_transcription"] else "No",
+            )
+            .add_field(
+                name="Archived", value="Yes" if self.submission["archived"] else "No"
+            )
+        )
+
+        if self.transcription:
+            embed.description = limit_str(self.transcription["text"], 200)
+        if self.submission.get("content_url"):
+            embed.set_image(url=self.submission["content_url"])
+        if self.submission.get("tor_url"):
+            embed.add_field(
+                name="ToR Post", value=f"[Link]({self.submission['tor_url']})"
+            )
+        if self.submission.get("url"):
+            embed.add_field(
+                name="Partner Post", value=f"[Link]({self.submission.get('url')})"
+            )
+        if tr_link:
+            embed.add_field(name="Transcription", value=tr_link)
+        subreddit = (
+            self.submission["url"].split("/")[4] if self.submission["url"] else None
+        )
+        if subreddit:
+            embed.set_author(
+                name=f"r/{subreddit}", url=f"https://reddit.com/r/{subreddit}"
+            )
+
+        return embed
+
+    def _clear(self) -> None:
+        """Reset the local scope caches for a new search."""
+        self.transcription: Union[Dict, None] = None
+        self.submission: Union[Dict, None] = None
+        self.author: Union[Dict, None] = None
```

Comment:
I wonder, what would happen if two users execute the `find` comment at the same time? Could they overwrite each other's caches, leading to corrupted data/wrong responses?

Reply:
...fuck. Yeah, I think you're right. I think this is the first time I've ever really worked with async code. I think I can fix that by keeping it in the local scope and not using a cache, but does that make it less clean I think. I'll play around with it.